src/components/node-list/node-list-items.js

```diff
@@ -331,3 +332,122 @@ export const getFilteredItems = createSelector(
     };
   }
 );
+
+// the following are selectors needed for the tree list data
+/**
+ * returns an array of the corresponding filtered nodes
+ * filtered nodes for each modular pipeline
+ */
+export const getFilteredModularPipelineNodes = createSelector(
+  [
+    getFilteredNodeItems,
+    getFilteredModularPipelineItems,
+    (state) => state.modularPipelineIds,
+    (state) => state.nodeTypeIDs,
+  ],
+  (
+    filteredNodeItems,
+    filteredModularPipelines,
+    modularPipelineIDs,
+    nodeTypeIDs
+  ) => {
+    const modularPipelineNodes = arrayToObject(modularPipelineIDs, () => []);
+    const { modularPipeline } = filteredModularPipelines;
+
+    // assumption: each node is unique and will only exist once on the flowchart, hence we are only taking
+    // the deepest nested modular pipeline as the node's modular pipeline
+    nodeTypeIDs.forEach((key) => {
+      filteredNodeItems[key]?.forEach((node, i) => {
+        if (node.modularPipelines.length > 1) {
+          node.modularPipelines = [
+            node.modularPipelines[node.modularPipelines.length - 1],
+          ];
+        }
+      });
+    });
+
+    // create a new field for the topmost / root pipeline
+    modularPipelineNodes['main'] = [];
+
+    // go through each type of nodes according to the order of specified node types in normalize-data
+    // first to identify root level nodes
+    nodeTypeIDs.forEach((key) => {
```

Comment:
I think prefer to use more specific names in these loops if the contents of `key` is known, so e.g. `nodeTypeID` here in that case?

Reply:
Thanks! I've incorporaed this in [this commit](https://github.com/quantumblacklabs/kedro-viz/pull/479/commits/a3ce810327dab1d6e78ddf9cc07f21411279e63e)